spec/formatting.md

```diff
@@ -18,3 +18,236 @@ the local variable takes precedence.
+
+During selection, an Expression handler must only emit Resolution and Selection errors.
+During formatting, an Expression handler must only emit Resolution and Formatting errors.
+
+In all cases, when encountering an error,
+a message formatter must provide some representation of the message.
+An informative error or errors must also be separately provided.
+When a message contains more than one error,
+or contains some error which leads to further errors,
+an implementation which does not emit all of the errors
+should prioritise Syntax and Data Model errors over others.
+
+When an error occurs in the resolution of an Expression or Markup Option,
+the Expression or Markup in question is processed as if the option were not defined.
+This may allow for the fallback handling described below to be avoided,
+though an error must still be emitted.
+
+When an error occurs within a Selector,
+the selector must not match any VariantKey other than the catch-all `*`
+and a Resolution or Selector error is emitted.
+
+## Fallback String Representations
+
+The formatted string representation of a message with a Syntax or Data Model error
+is the concatenation of U+007B LEFT CURLY BRACKET `{`,
+a fallback string,
+and U+007D RIGHT CURLY BRACKET `}`.
+If a fallback string is not defined,
+the U+FFFD REPLACEMENT CHARACTER `�` character is used,
+resulting in the string `{�}`.
+
+When an error occurs in a Placeholder that is being formatted,
+the fallback string representation of the Placeholder
+always starts with U+007B LEFT CURLY BRACKET `{`
+and ends with U+007D RIGHT CURLY BRACKET `}`.
+Between the brackets, the following contents are used:
+
+- Expression with Literal Operand: U+0028 LEFT PARENTHESIS `(`
+  followed by the value of the Literal,
+  and then by U+0029 RIGHT PARENTHESIS `)`
+
+  Examples: `{(horse)}`, `{(42)}`
+
+- Expression with Variable Operand: U+0024 DOLLAR SIGN `$`
+  followed by the Variable Name of the Operand
+
+  Example: `{$user}`
+
+- Expression with no Operand: U+003A COLON `:` followed by the Expression Name
+
+  Example: `{:platform}`
+
+- Markup start: U+002B PLUS SIGN `+` followed by the MarkupStart Name
+
+  Example: `{+tag}`
+
+- Markup end: U+002D HYPHEN-MINUS `-` followed by the MarkupEnd Name
+
+  Example: `{-tag}`
+
+- Otherwise: The U+FFFD REPLACEMENT CHARACTER `�` character
+
+  Example: `{�}`
+
+Option names and values are not included in the fallback string representations.
+
+When an error occurs in an Expression with a Variable Operand
+and the Variable refers to a local variable Declaration,
+the fallback string is formatted based on the Expression of the Declaration,
+rather than the Expression of the Placeholder.
+
+For example, attempting to format either of the following messages within a context that
+does not provide for the function `:func` to be successfully resolved:
+
+```
+let $var = {(horse) :func}
+{The value is {$var}.}
+```
+
+```
+let $var = {(horse)}
```

Comment:
Nitpick: I don't think the second example should be valid
```
let $var = {(horse)}
{The value is {$var :func}.}
```
Because the whole point of local variables would be reuse, mostly for consistency.
If change the function in the message body, the whole consistency goes out the windows.

Yes, not difficult to implement (just a bit worse performance).
But kind of no good use case, with potential of misuse.

So my suggestion would be to remove the second example.

Reply:
This is meant as a minimal example showing that the identifiers of local variables should never end up being included in the formatted output. It's not really meant to be a useful message, but it's technically valid.